### ISSUE TITLE
[Feature] Click the "Image" button on the menu bar to upload images

### ISSUE THREAD

**Issue by LittleSound** [Saul-Mirone/milkdown#635]
I found that there is already a plugin called `@milkdown/plugin-upload` that supports image uploads, but it only supports dragging and dropping or pasting to trigger the upload, and clicking the "image" button in the menu bar still asks the user to fill in the link.
Is it possible to integrate the functionality of `@milkdown/plugin-upload` into the menu bar?

**Comment by Saul-Mirone** [comment]
PR welcome :)

**Comment by LittleSound** [comment]
我可以开发这个功能，但是我想知道要如何开始。
假设预期的效果是在安装 '@milkdown/plugin-upload' 插件后，图片组件的 `Input Image Link` 旁边增加一个文件上传按钮，点击后打开文件选择器，我需要如何实现呢？
<img width="1239" alt="image" src="https://user-images.githubusercontent.com/19204772/180931103-d1c5e005-5e68-4ef1-9c45-0a02e5c16f44.png">

**Comment by LittleSound** [comment]
@Saul-Mirone I tried to read the relevant code.

This seems to require some modification to the base component, as neither the `image` nor the `input-chip` components provide `slots` that can append a button, nor does it provide the ability to expand a button to `<input type="file" >` if the button in `menu` is modified directly.

Not to mention that I don't know how to change a `view` in another plugin that is already rendered by another plugin, which leaves me with no way to do so.

Maybe I need to do a direct `document` operation to append an `<input type="file" >` button?

Can you give me some advice on how to implement this feature from the author's point of view? Thank you.

**Comment by LittleSound** [comment]
据我了解，文件选择需要通过 `<input type="file" >` 触发，所以至少需要在用户点击的按钮旁放置一个  `<input type="file" >`，这似乎没办法直接配置 `@milkdown/plugin-menu` 或者命令来实现。

**Comment by Saul-Mirone** [comment]
Maybe not, it's also possible to use a hidden input and click it programmatically.

**Comment by LittleSound** [comment]
> Maybe not, it's also possible to use a hidden input and click it programmatically.

I know it can be triggered programmatically, but the location of this `<input type="file" >` has to be the same as the location the user clicks on, because in IOS the file selector opens a dropdown menu at the location of `<input type="file" >`.

> Here is an example (Twitter):
> <img width="914" alt="image" src="https://user-images.githubusercontent.com/19204772/181180354-474e4fd1-7bed-4ce2-bd6e-590a86b145fa.png">
> ![image](https://user-images.githubusercontent.com/19204772/181180180-f326a574-356a-4666-bae9-a5b432ab1c2f.png)

**Comment by LittleSound** [comment]
First of all, this sounds like great news!

But, if I still want to be able to add file upload functionality to the `Image` component, is there any way to do that?

It seems to me that adding a `! []()` string in the `milkdown` editor in a variety of ways (direct input, slash command, menu on the left side of the line), it seems unreasonable that the only way to use the upload process would be to click the menu bar button, which would fall into the restriction that the only way to upload is to copy and paste; whereas the `Image` component is the way to go on the image path, and it makes more sense to add local file upload functionality here.


首先，这听起来是个好消息！

但是，如果我还是希望可以在 `Image` 组件中添加文件上传功能的话，是否有什么可行的方式呢？

在我看来在 `milkdown` 编辑器中添加一个 `![]()` 字符串的方式有很多种 (直接输入，斜杠命令，行左侧的菜单)，如果只有点击菜单栏按钮才能使用上传流程，显得很不合理，这就又陷入了只有复制粘贴才能上传类似的限制中；而 `Image` 组件是图片功路径上的必进之路，在此添加本地文件上传功能更为合理。

**Comment by Saul-Mirone** [comment]
I think you may need to provide another image [node view](https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews). Current implementation is in [theme-pack-helper](https://github.com/Saul-Mirone/milkdown/blob/main/packages/theme-pack-helper/src/renderer-preset/image.ts). You may want to create a new one.

**Comment by LittleSound** [comment]
Thanks. Are there any examples of how to override an existing `image [node view](https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews).` in my plugin?

**Comment by Saul-Mirone** [comment]
You can just provide your view and users can extend the current theme to apply it:

```typescript
import { nord } from '@milkdown/theme-nord'

const myImageView = (emotion, manager) => {
  manager.set<ThemeImageType>('image', () => {
    // your implementation here
  });
}

nord.override((emotion, manager) => {
  myImageView(emotion, manager);
});
```

You can also provide a helper function to make users can use it simpler.

**Comment by LittleSound** [comment]
Thank you very much.

**Comment by Saul-Mirone** [comment]
I can add your plugin in the list of https://github.com/Saul-Mirone/awesome-milkdown once your plugin complete.

**Comment by LittleSound** [comment]
I finished the plugin:
[milkdown-plugin-image-picker](https://github.com/LittleSound/milkdown-plugin-image-picker)

This took me about two weekends of, It may still have a lot of room for improvement, but it's already working

**Comment by Saul-Mirone** [comment]
https://github.com/Saul-Mirone/awesome-milkdown/commit/5d81fa5920cbce37f45bfe856db097856e408ae5

**Comment by kinecology** [comment]
I'm using this `https://milkdown.dev/docs/api/plugin-upload` with my own uploader:

```
  import { uploader } from './milkdownUploadHandler';
// ...
 ctx.update(uploadConfig.key, (prev) => ({
                        ...prev,
                        uploader,
                        }));
```
and it's working fine for drop or paste.

Now I would like to add a toolbar button to do uploads, and I have that working elsewhere (in a view that does not involve milkdown). So, I thought it would be easy to integrate them but I am having trouple figuring out **how to pass a command to the editor or plugin handler to invoke it programmatically**?

In other words, I have code to capture or select an image at get the file, but then how can I pass the image file to invoke `plugin-upload` at the current cursor position?

I thought it might be as simple as,
```
    let saveCapturedImage = async () => {

      theEditor.applyCommand(uploadCommand, [photoFile]);
  
      sheetOpenCamera = false;
      photoInput.value = "";
    };
```

..., but I can't find a command or logic. Perhaps I have to use this [second image picker plugin](https://github.com/LittleSound/milkdown-plugin-image-picker)?

Thanks for advice.
